src/test/util_string_tests.cpp

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2024-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <util/string.h>
+
+#include <boost/test/unit_test.hpp>
+
+using namespace util;
+
+BOOST_AUTO_TEST_SUITE(util_string_tests)
+
+BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
+{
+    // Compile-time sanity checks
+    static_assert([] {
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%%");
+        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("_%");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%_");
+        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_%");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("_%%");
+        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%%%_");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%_%");
+        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_%%");
+        ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("_%%%");
+        return true; // All checks above compiled and passed
+    }());
```

Comment:
the errors are really ugly this way:
> util_string_tests.cpp:16:19: error: static assertion expression is not an integral constant expression
    static_assert([] {
                  ^~~~

Would there be any disadvantage in inlining them?
```C++
BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
{
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("");
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%");
...
    ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_%%");
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("_%%%");
```

Reply:
While the code is slightly simpler (-4 lines, indentation), the error message in case of failure is also simpler:

<details>
<summary>with static assert</summary>

```
bitcoin/src/test/util_string_tests.cpp:16:19: error: static assertion expression is not an integral constant expression
    static_assert([] {
                  ^~~~
bitcoin/src/util/string.h:39:34: note: subexpression not valid in a constant expression
        if (num_params != count) throw "Format specifier count must match the argument count!";
                                 ^
bitcoin/src/test/util_string_tests.cpp:22:9: note: in call to 'Detail_CheckNumFormatSpecifiers({&"% % %_"[0], 6})'
        ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("% % %_");
        ^
bitcoin/src/test/util_string_tests.cpp:16:19: note: in call to '&[] {
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("");
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%");
    ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_");
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("%%_");
    ConstevalFormatString<0>::Detail_CheckNumFormatSpecifiers("_%%");
    ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("% % %_");
    ConstevalFormatString<1>::Detail_CheckNumFormatSpecifiers("%_%%");
    return true;
}->operator()()'
    static_assert([] {
                  ^
1 error generated.
make[2]: *** [src/test/CMakeFiles/test_bitcoin.dir/util_string_tests.cpp.o] Error 1
make[1]: *** [src/test/CMakeFiles/test_bitcoin.dir/all] Error 2
make: *** [all] Error 2
```

</details>

<details>
<summary>without static assert</summary>

```
unknown location:0: fatal error: in "util_string_tests/ConstevalFormatString_NumSpec": C string: Format specifier count must match the argument count!
src/test/util_string_tests.cpp:13: last checkpoint: "ConstevalFormatString_NumSpec" test entry

*** 1 failure is detected in the test module "Bitcoin Core Test Suite"
```

</details>

> Moving the checks from being done at compile-time to be done at runtime means that possible errors will be caught later

In reality these lines will rarely be modified - but probably read more often, so I'd go for the simpler code.
If you still insist that this is better, please resolve the comment.